cmd/atlas/internal/cmdext/cmdext.go

```diff
@@ -192,3 +201,63 @@ func blockError(name string, b *hclsyntax.Block) func(string, ...any) error {
 		return fmt.Errorf("%s.%s: %w", name, b.Labels[1], fmt.Errorf(format, args...))
 	}
 }
+
+// StateReader returns the state reader for the given scheme.
+func StateReader(scheme string) (ReadStateFunc, bool) {
+	r, ok := states[scheme]
+	return r, ok
+}
+
+type (
+	// ReadStateOptions for external state readers.
+	ReadStateOptions struct {
+		URLs []*url.URL        // URL of the resource.
+		Dev  *sqlclient.Client // Client for the dev database.
+	}
+	// ReadStateFunc is the function signature for external state readers.
+	ReadStateFunc func(context.Context, *ReadStateOptions) (migrate.StateReader, error)
+)
+
+// states is a global registry for external state readers. The key is the
+// URL scheme and the value is the function that returns the StateReader.
+var states = map[string]ReadStateFunc{
+	"ent": EntSchemaState,
+}
+
+// EntSchemaState returns a migrate.StateReader that reads the schema from an ent.Schema.
+func EntSchemaState(ctx context.Context, opts *ReadStateOptions) (migrate.StateReader, error) {
+	if opts.Dev == nil || len(opts.URLs) != 1 {
+		return nil, errors.New("schema url and dev database are required")
+	}
+	path := filepath.Join(opts.URLs[0].Host, opts.URLs[0].Path)
+	graph, err := entc.LoadGraph(path, &gen.Config{})
+	if err != nil {
+		return nil, fmt.Errorf("loading schema %s: %w", path, err)
+	}
+	m, err := entschema.NewMigrate(
```

Comment:
Why not use `NewMigrateURL()`? You can pass in the URL from the args directly.

Reply:
I want to reuse the same database connection.